File: src/rpctests/graphql.py
#!/usr/bin/python3
"""
This script connects to an Ethereum node via HTTP and sends requests via GraphQL.
"""

import aiohttp
import argparse
import asyncio
import json
import logging
import pathlib
import shutil
import signal
import sys
import tempfile
import urllib.parse

from .common import graphql

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)


def parse_github_tree_url(url: str) -> tuple[str, str, str, str]:
    """Helper to parse the GitHub URL: extracts owner, repo, branch, and path from a GitHub tree URL."""
    try:
        parts = urllib.parse.urlparse(url).path.strip('/').split('/')
        if len(parts) < 5 or parts[2] != 'tree':
            raise ValueError(f"Invalid GitHub 'tree' URL format: {url}.")

        owner, repo, _, branch, *path_parts = parts
        folder_path = '/'.join(path_parts)
        return owner, repo, branch, folder_path
    except Exception as e:
        raise ValueError(f"Failed to parse GitHub URL: {url}. Error: {e}")


async def download_github_directory(url: str) -> pathlib.Path:
    """Downloads all files from specified GitHub URL into a temporary local folder using GitHub REST API."""
    owner, repo, branch, folder_path = parse_github_tree_url(url)
    api_url = f"https://api.github.com/repos/{owner}/{repo}/contents/{folder_path}?ref={branch}"

    try:
        # Create a secure, temporary directory
        temp_dir = pathlib.Path(tempfile.mkdtemp())
        logger.info(f"⬇️ Downloading test files to temporary directory: {temp_dir}")

        # Use aiohttp for async API calls
        async with aiohttp.ClientSession() as session:
            # Request directory contents
            async with session.get(api_url) as response:
                if response.status != aiohttp.http.HTTPStatus.OK:
                    content = await response.text()
                    raise Exception(f"GitHub API error {response.status}. URL: {api_url}. Response: {content[:100]}...")

                # The API returns a list of files/directories
                contents = await response.json()

            download_tasks = []
            for item in contents:
                # Test files are expected to be plain JSON files
                if item['type'] == 'file' and item['name'].endswith('.json'):
                    # Get the raw download URL
                    download_url = item['download_url']
                    file_path = temp_dir / item['name']

                    # Create a task to download each file concurrently
                    task = download_single_file(session, download_url, file_path)
                    download_tasks.append(task)

            # Wait for all file downloads to complete
            await asyncio.gather(*download_tasks)
            logger.info(f"✅ Downloaded {len(download_tasks)} test files.")
            return temp_dir
    except Exception as e:
        logger.error(f"❌ Error during GitHub download: {e}")
        raise e


async def download_single_file(session: aiohttp.ClientSession, url: str, path: pathlib.Path):
    """Downloads a single file asynchronously and saves it to the specified path."""
    async with session.get(url) as response:
        response.raise_for_status()  # Raises an exception for bad status codes (4xx or 5xx)
        content = await response.read()
        path.write_bytes(content)
        logger.debug(f"Downloaded {path.name}")


async def execute_tests(processor: graphql.QueryProcessor, url: str, stop_at_error: bool, test_number: int | None):
    """Downloads files from the specified URL, discovers test files, and executes them."""
    temp_dir = None
    try:
        # Download the GitHub directory contents into a temporary folder
        temp_dir = await download_github_directory(url)

        logger.info(f"Starting test execution using files from {temp_dir}")

        # Discover test files (only looking for .json files) and sort them by name
        test_files = list(temp_dir.glob('*.json'))
        if not test_files:
            logger.warning(f"⚠️ No *.json files found in {temp_dir}. Aborting tests.")
            return 1
        test_files = sorted(test_files)

        total_tests = len(test_files) if not test_number else 1
        passed_tests = 0

        # Iterate and execute all tests by establishing and reusing the same session
        async with processor as session:
            for i, test_file_path in enumerate(test_files):
                if test_number and test_number != i:
                    continue
                test_name = test_file_path.name
                try:
                    test_case = json.loads(test_file_path.read_text(encoding='utf-8'))

                    query = test_case.get("request", "").strip()
                    if not query:
                        logger.error(f"❌ Test {i + 1} FAILED: 'request' field is missing in {test_name}.")
                        continue
                    expected_responses = test_case.get("responses", [])
                    if not expected_responses:
                        logger.error(f"❌ Test {i + 1} FAILED: 'responses' field is missing in {test_name}.")
                        continue
                except json.JSONDecodeError as e:
                    logger.error(f"❌ Test {i + 1} FAILED: Invalid JSON format in {test_name}: {e}")
                    continue

                # Execute the query
                actual_result = await session.execute(query)

                # Compare actual vs expected: the test passes if the actual result matches ANY of the expected responses
                is_passing = False
                actual_data, actual_errors = actual_result.data, actual_result.errors
                expected_data, expected_errors = None, None
                for expected in expected_responses:
                    # Assume the expected response contains { "data": { ... } } or { "errors": { ... } } or both
                    expected_data = expected.get("data")
                    if actual_data == expected_data:
                        is_passing = True
                        break  # Found a match, test passes
                    expected_errors = expected.get("errors")
                    if expected_errors and actual_errors:
                        is_passing = True
                        break  # Found a match, test passes

                # Log result and update counters
                if is_passing:
                    passed_tests += 1
                    logger.info(f"✅ Test {i + 1} {test_name} PASSED.")
                else:
                    logger.info(f"❌ Test {i + 1} {test_name} FAILED: actual result didn't match any expected response.")
                    logger.warning(f"Request: {query}")
                    if expected_data:
                        logger.warning(f"Expected data: {expected_data}")
                        if actual_data:
                            logger.warning(f"Actual data:   {actual_data}")
                        else:
                            logger.warning(f"Actual error:  {actual_errors}")
                    else:
                        logger.warning(f"Expected error: {expected_errors}")
                        if actual_errors:
                            logger.warning(f"Actual error:  {actual_errors}")
                        else:
                            logger.warning(f"Actual data:   {actual_data}")
                    if stop_at_error:
                        logger.info(f"Testing finished after first error. Passed: {passed_tests}/{total_tests}")
                        return 1

            logger.info(f"Testing finished. Passed: {passed_tests}/{total_tests}")
            status = 0 if passed_tests == total_tests else 1
            return status

    except (ValueError, Exception) as e:
        logger.error(f"❌ Test setup/execution failed: {e}")
        return 1

    finally:
        # Clean up the temporary directory created by the download function
        if temp_dir and temp_dir.exists():
            logger.info(f"Cleaning up temporary directory: {temp_dir}")
            try:
                shutil.rmtree(temp_dir)
            except OSError as e:
                logger.error(f"Failed to remove temporary directory {temp_dir}: {e}")


async def main():
    """Main function to run the GraphQL processor."""

    # Parse command-line arguments
    parser = argparse.ArgumentParser(
        description="Queries an Ethereum node via GraphQL."
    )
    parser.add_argument(
        "--http_url",
        type=str,
        default="http://127.0.0.1:8545/graphql",
        help="The GraphQL URL of the Ethereum node (default: http://127.0.0.1:8545/graphql)",
    )
    parser.add_argument(
        "--stop_at_first_error",
        action="store_true",
        default=False,
        help="Flag indicating that execution must be stopped at first error",
    )
    parser.add_argument(
        "--test_number",
        type=int,
        default=None,
        help="The progressive number of the test file to execute. If specified, only such test gets executed",
    )
    # User MUST provide EITHER --query OR --tests_url, but NOT BOTH.
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument(
        "--query",
        type=str,
        default=None,
        help="The GraphQL query as string",
    )
    group.add_argument(
        "--tests_url",
        type=str,
        default=None,
        help="The URL or path indicating where to read test files",
    )
    args = parser.parse_args()

    # Create the GraphQL query processor
    processor = graphql.QueryProcessor(args.http_url)

    # Setup signal handler for graceful shutdown
    async def signal_handler():
        print("")
        logger.info("🏁 Received interrupt signal (Ctrl+C)")
        await processor.close()

    loop = asyncio.get_running_loop()
    for sig in [signal.SIGINT, signal.SIGTERM]:
        loop.add_signal_handler(sig, lambda: asyncio.create_task(signal_handler()))

    status = 0
    try:
        if args.query:
            result = await processor.execute_async(args.query)
            logger.info(f"✅ Result: {result}")
        else:
            status = await execute_tests(processor, args.tests_url, args.stop_at_first_error, args.test_number)
    except KeyboardInterrupt:
        logger.info("Interrupted by user")
    except Exception as e:
        logger.error(f"❌ Unexpected error: {e}")
        status = e
    finally:
        sys.exit(status)


if __name__ == "__main__":
    """ Usage: python graphql.py """
    asyncio.run(main())

File: internal/tools/graphql.go
package tools

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/urfave/cli/v2"
)

var graphqlCommand = &cli.Command{
	Name:  "graphql",
	Usage: "Execute GraphQL queries against an Ethereum node",
	Flags: []cli.Flag{
		&cli.StringFlag{
			Name:  "http-url",
			Value: "http://127.0.0.1:8545/graphql",
			Usage: "GraphQL URL of the Ethereum node",
		},
		&cli.StringFlag{
			Name:  "query",
			Usage: "GraphQL query string (mutually exclusive with --tests-url)",
		},
		&cli.StringFlag{
			Name:  "tests-url",
			Usage: "GitHub tree URL with test files (mutually exclusive with --query)",
		},
		&cli.BoolFlag{
			Name:  "stop-at-first-error",
			Usage: "Stop execution at first test error",
		},
		&cli.IntFlag{
			Name:  "test-number",
			Value: -1,
			Usage: "Run only the test at this index (0-based)",
		},
	},
	Action: runGraphQL,
}

type graphqlTestCase struct {
	Request   string            `json:"request"`
	Responses []json.RawMessage `json:"responses"`
}

func runGraphQL(c *cli.Context) error {
	httpURL := c.String("http-url")
	query := c.String("query")
	testsURL := c.String("tests-url")
	stopAtError := c.Bool("stop-at-first-error")
	testNumber := c.Int("test-number")

	if query == "" && testsURL == "" {
		return fmt.Errorf("must specify either --query or --tests-url")
	}
	if query != "" && testsURL != "" {
		return fmt.Errorf("--query and --tests-url are mutually exclusive")
	}

	client := &http.Client{}

	if query != "" {
		result, err := executeGraphQLQuery(client, httpURL, query)
		if err != nil {
			return err
		}
		log.Printf("Result: %s", result)
		return nil
	}

	return executeGraphQLTests(client, httpURL, testsURL, stopAtError, testNumber)
}

func executeGraphQLQuery(client *http.Client, url, query string) ([]byte, error) {
	payload, err := json.Marshal(map[string]string{"query": query})
	if err != nil {
		return nil, fmt.Errorf("marshal query: %w", err)
	}
	req, err := http.NewRequest("POST", url, bytes.NewReader(payload))
	if err != nil {
		return nil, fmt.Errorf("create request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("execute query: %w", err)
	}
	defer resp.Body.Close()

	return io.ReadAll(resp.Body)
}

func executeGraphQLTests(client *http.Client, httpURL, testsURL string, stopAtError bool, testNumber int) error {
	// Download test files from GitHub
	tempDir, err := downloadGitHubDirectory(client, testsURL)
	if err != nil {
		return fmt.Errorf("download tests: %w", err)
	}
	defer func() {
		log.Printf("Cleaning up temporary directory: %s", tempDir)
		_ = os.RemoveAll(tempDir)
	}()

	log.Printf("Starting test execution using files from %s", tempDir)

	// Discover and sort test files
	entries, err := os.ReadDir(tempDir)
	if err != nil {
		return fmt.Errorf("read test dir: %w", err)
	}

	var testFiles []string
	for _, e := range entries {
		if !e.IsDir() && strings.HasSuffix(e.Name(), ".json") {
			testFiles = append(testFiles, e.Name())
		}
	}
	sort.Strings(testFiles)

	if len(testFiles) == 0 {
		log.Printf("Warning: no *.json files found in %s. Aborting tests.", tempDir)
		return fmt.Errorf("no test files found")
	}

	totalTests := len(testFiles)
	if testNumber >= 0 {
		totalTests = 1
	}
	passedTests := 0

	graphqlClient := &http.Client{}

	for i, testFile := range testFiles {
		if testNumber >= 0 && testNumber != i {
			continue
		}

		testPath := filepath.Join(tempDir, testFile)
		data, err := os.ReadFile(testPath)
		if err != nil {
			log.Printf("Test %d FAILED: cannot read %s: %v", i+1, testFile, err)
			continue
		}

		var tc graphqlTestCase
		if err := json.Unmarshal(data, &tc); err != nil {
			log.Printf("Test %d FAILED: invalid JSON in %s: %v", i+1, testFile, err)
			continue
		}

		if tc.Request == "" {
			log.Printf("Test %d FAILED: 'request' field is missing in %s", i+1, testFile)
			continue
		}
		if len(tc.Responses) == 0 {
			log.Printf("Test %d FAILED: 'responses' field is missing in %s", i+1, testFile)
			continue
		}

		// Execute query
		actualResult, err := executeGraphQLQuery(graphqlClient, httpURL, strings.TrimSpace(tc.Request))
		if err != nil {
			log.Printf("Test %d FAILED: query execution error: %v", i+1, err)
			if stopAtError {
				log.Printf("Testing finished after first error. Passed: %d/%d", passedTests, totalTests)
				return fmt.Errorf("stopped at first error")
			}
			continue
		}

		// Parse actual result
		var actualData map[string]any
		if err := json.Unmarshal(actualResult, &actualData); err != nil {
			log.Printf("Test %d FAILED: cannot parse response: %v", i+1, err)
			continue
		}

		// Compare actual vs expected: test passes if actual matches ANY expected response
		passing := false
		for _, expectedRaw := range tc.Responses {
			var expected map[string]any
			if err := json.Unmarshal(expectedRaw, &expected); err != nil {
				continue
			}

			// Check if actual data matches expected data
			actualDataField := actualData["data"]
			expectedDataField := expected["data"]
			if jsonEqual(actualDataField, expectedDataField) {
				passing = true
				break
			}

			// Check if both have errors
			if expected["errors"] != nil && actualData["errors"] != nil {
				passing = true
				break
			}
		}

		if passing {
			passedTests++
			log.Printf("Test %d %s PASSED.", i+1, testFile)
		} else {
			log.Printf("Test %d %s FAILED: actual result didn't match any expected response.", i+1, testFile)
			log.Printf("Request: %s", strings.TrimSpace(tc.Request))
			log.Printf("Actual:  %s", string(actualResult))
			if stopAtError {
				log.Printf("Testing finished after first error. Passed: %d/%d", passedTests, totalTests)
				return fmt.Errorf("stopped at first error")
			}
		}
	}

	log.Printf("Testing finished. Passed: %d/%d", passedTests, totalTests)
	if passedTests != totalTests {
		return fmt.Errorf("some tests failed: %d/%d passed", passedTests, totalTests)
	}
	return nil
}

func jsonEqual(a, b any) bool {
	aj, err1 := json.Marshal(a)
	bj, err2 := json.Marshal(b)
	if err1 != nil || err2 != nil {
		return false
	}
	return bytes.Equal(aj, bj)
}

type githubContent struct {
	Name        string `json:"name"`
	Type        string `json:"type"`
	DownloadURL string `json:"download_url"`
}

func parseGitHubTreeURL(rawURL string) (owner, repo, branch, folderPath string, err error) {
	u, err := url.Parse(rawURL)
	if err != nil {
		return "", "", "", "", fmt.Errorf("parse URL: %w", err)
	}
	parts := strings.Split(strings.Trim(u.Path, "/"), "/")
	if len(parts) < 5 || parts[2] != "tree" {
		return "", "", "", "", fmt.Errorf("invalid GitHub tree URL format: %s", rawURL)
	}
	owner = parts[0]
	repo = parts[1]
	branch = parts[3]
	folderPath = strings.Join(parts[4:], "/")
	return
}

func downloadGitHubDirectory(client *http.Client, treeURL string) (string, error) {
	owner, repo, branch, folderPath, err := parseGitHubTreeURL(treeURL)
	if err != nil {
		return "", err
	}

	apiURL := fmt.Sprintf("https://api.github.com/repos/%s/%s/contents/%s?ref=%s", owner, repo, folderPath, branch)

	tempDir, err := os.MkdirTemp("", "graphql-tests-*")
	if err != nil {
		return "", fmt.Errorf("create temp dir: %w", err)
	}
	log.Printf("Downloading test files to temporary directory: %s", tempDir)

	resp, err := client.Get(apiURL)
	if err != nil {
		_ = os.RemoveAll(tempDir)
		return "", fmt.Errorf("fetch GitHub API: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		_ = os.RemoveAll(tempDir)
		body, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("GitHub API error %d: %s", resp.StatusCode, string(body[:min(len(body), 100)]))
	}

	var contents []githubContent
	if err := json.NewDecoder(resp.Body).Decode(&contents); err != nil {
		_ = os.RemoveAll(tempDir)
		return "", fmt.Errorf("decode GitHub API response: %w", err)
	}

	downloaded := 0
	for _, item := range contents {
		if item.Type != "file" || !strings.HasSuffix(item.Name, ".json") {
			continue
		}

		fileResp, err := client.Get(item.DownloadURL)
		if err != nil {
			log.Printf("Warning: failed to download %s: %v", item.Name, err)
			continue
		}

		data, err := io.ReadAll(fileResp.Body)
		fileResp.Body.Close()
		if err != nil {
			log.Printf("Warning: failed to read %s: %v", item.Name, err)
			continue
		}

		if err := os.WriteFile(filepath.Join(tempDir, item.Name), data, 0644); err != nil {
			log.Printf("Warning: failed to write %s: %v", item.Name, err)
			continue
		}
		downloaded++
	}

	log.Printf("Downloaded %d test files.", downloaded)
	return tempDir, nil
}
